src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3291,30 +3398,147 @@ void MacroAssembler::lookup_secondary_supers_table(Register r_sub_klass,
+    r_array_index = temp3,
+    slot          = temp4, // NOTE: "slot" can't be Z_R0 otherwise z_sllg and z_rllg instructions below will mess up!!!!
+    r_bitmap      = Z_R1_scratch;
+
+  z_llgc(slot, Address(r_super_klass, Klass::hash_slot_offset()));
+
+  // Initialize r_result with 0 (indicating success). If searching fails, r_result will be loaded
+  // with 1 (failure) at the end of this method.
+  clear_reg(result, true /* whole_reg */, false /* set_cc */); // result = 0
+
+  z_lg(r_bitmap, Address(r_sub_klass, Klass::secondary_supers_bitmap_offset()));
+
+  // First check the bitmap to see if super_klass might be present. If
+  // the bit is zero, we are certain that super_klass is not one of
+  // the secondary supers.
+#ifdef ASSERT
+  z_lgr(Z_R0, slot);
+#endif // ASSERT
+  z_xilf(slot, (u1)(Klass::SECONDARY_SUPERS_TABLE_SIZE - 1)); // slot ^ 63 === 63 - slot (mod 64)
+  z_sllg(r_array_index, r_bitmap, /*d2 = */ 0, /* b2 = */ slot);
+
+  testbit(r_array_index, Klass::SECONDARY_SUPERS_TABLE_SIZE - 1);
+  branch_optimized(bcondAllZero, L_failure);
+
+  const Register
+    r_array_base   = temp1,
+    r_array_length = temp2;
+
+  // Get the first array index that can contain super_klass into r_array_index.
+  // NOTE: Z_R1_scratch is holding bitmap (look above for r_bitmap). So let's try to save it.
+  //       On the other hand, r_array_base/temp1 is free at current moment (look at the load operation below).
+  pop_count_long(r_array_index, r_array_index, temp1); // kills r_array_base/temp1 on machines older than z15
+
+  // The value i in r_array_index is >= 1, so even though r_array_base
+  // points to the length, we don't need to adjust it to point to the data.
+  assert(Array<Klass*>::base_offset_in_bytes() == wordSize, "Adjust this code");
+  assert(Array<Klass*>::length_offset_in_bytes() == 0, "Adjust this code");
+
+  // We will consult the secondary-super array.
+  z_lg(r_array_base, Address(r_sub_klass, in_bytes(Klass::secondary_supers_offset())));
+
+  // NB! r_array_index is off by 1. It is compensated by keeping r_array_base off by 1 word.
+  z_sllg(r_array_index, r_array_index, LogBytesPerWord); // scale, r_array_index is loaded by popcnt above
+
+  z_cg(r_super_klass, Address(r_array_base, r_array_index));
+  branch_optimized(bcondEqual, L_done); // found a match
+
+  // restore the value of slot
+  z_xilf(slot, (u1)(Klass::SECONDARY_SUPERS_TABLE_SIZE - 1)); // slot ^ 63 === 63 - slot (mod 64)
+#ifdef ASSERT
+  NearLabel ok;
+  z_cgr(slot, Z_R0);
+  z_bre(ok);
+  stop("slot value got corrupted");
+  bind(ok);
+#endif // ASSERT
+
+  // Linear probe. Rotate the bitmap so that the next bit to test is
+  // in Bit 1.
+
+  // below, we just want to subtract slot from 64, because we don't have a rotate right instruction.
+  // Oh, also we don't have instruction to do : 64 - slot
+  // So, negate slot and add 64 to it. i.e. 64 + (-slot)
+  z_lngr(slot, slot); // slot = -slot
+  z_aghi(slot, 64); // slot = slot + 64
+
```

Comment:
So, I'm trying to make sense of this.

First, you do

```
   slot = 63 - slot
   slot = 64 - slot

```
Why not instead reduce
`   slot = 64 - (63 - slot)`
which is
`   slot = slot - 63 + 64`
which is
`   slot = slot + 1`

```
jshell> void p() {
   ...>     for (int s = 0; s < 64; s++) {
   ...>         int n1 = s;
   ...>         n1 = 63 - n1;
   ...>         n1 = 64 - n1;
   ...>         int n2 = s + 1;
   ...>         if (n1 != n2) {
   ...>             throw new RuntimeException();
   ...>         }
   ...>     }
   ...> }
|  created method p()

jshell> p()


```

Reply:
I must say, that was a great suggestion. Thanks for pointing it out. I have pushed the change, please look at the new code :)